Initial development

kcmr opened the pull request:
Ember CLI addon to bundle Web Components using ES modules:

- Generates a bundle for each entry point path.
- Supports sourcemaps.
- Triggers a reload when the whatched files change.
- Optionally generates two bundles for modern (ES modules support) and legacy browsers. 
- Allows to not write the script tags (intended for lazy or on demand load of the scripts).
- Includes some utilities in the bundle:
    - Prevents common [double registration error](https://github.com/Polymer/polymer-starter-kit/issues/1123#issuecomment-434724376)
    - ~Prevents Safari 10 downloading `<script nomodule>` ([Browser issues](https://jakearchibald.com/2017/es-modules-in-browsers/#browser-issues))~ <sup>[1](#footnote-1)</sup>
- custom-es5-adapter is not required when transpiling Web Components using classes to ES5.

<sup><a name="footnote-1">1</a></sup> **Regarding polyfills**: 

Finally the script to prevent Safari 10 downloading `nomodule` scripts is not included because the addon is assuming that the hosting application supports that browser. Additionally, when `autoImport` is set to `false`, the script had to be imported from the addon folder (It doesn't sound good), so it's better to let the user include and import that script only if he needs it. Also, the file was included even if `modules` option was not enabled, so it was totally wrong.

The `customelements-define.js` script is always included in the bundle (not in vendor) because we know that it fixes a common issue when a script that registers a `customElements` is required by multiple files (very likely to happen). That script will be included in each of the generated bundles, but it is small enough (12 lines without minifying it) to not be an issue.

kcmr committed:
chore(package): update dependencies

kcmr committed:
feat: initial development

kcmr committed:
chore(package): remove addon directory from linted dirs

kcmr committed:
chore(package): add eslint

kcmr committed:
chore(package): change lint config settings

kcmr committed:
chore(package): change lint config settings

kcmr committed:
chore(package): fix fixable lint warnings

kcmr committed:
style(ember-cli-build): lint

kcmr committed:
chore(template-lint): change settings

kcmr committed:
chore: remove unused folders

kcmr committed:
chore(package): fix lint config

kcmr committed:
style: lint files

kcmr committed:
refactor: remove references to LitElement

kcmr committed:
chore(package): remove some devDependencies

kcmr committed:
chore(dummy): delete unused folders

kcmr committed:
test(config): change coverage reports folder for ember tests

kcmr committed:
test: add an acceptance test

kcmr committed:
refactor: prevent rootURL to be set from user options

kcmr committed:
refactor: change order of methods

kcmr committed:
chore: add jest config

kcmr committed:
test: back to nyc and mocha instead of jest

kcmr committed:
fix(babel): add required plugin for babel for legacy config

kcmr committed:
test: add coverage config for nyc

kcmr committed:
test: inserts script tag in index

kcmr committed:
refactor(rollup config): remove default values in params

kcmr committed:
test: output using modules

kcmr committed:
refactor: reduce required coverage and test branches

kcmr committed:
test: use yeoman-assert

kcmr committed:
test: addon without autoImport option

kcmr committed:
test: generates one bundle for each entrypointPath

kcmr committed:
chore(package): add missing command

kcmr committed:
chore(package): add description and keywords

kcmr committed:
docs(readme): add docs

kcmr committed:
chore(package): use files field in package instead of .npmignore

kcmr committed:
docs(readme): add badges and improve docs

kcmr committed:
chore(dummy): improve dummy app styles

kcmr committed:
docs(readme): improve docs

kcmr committed:
ci: add deploy config

kcmr committed:
chore: remove unused file

kcmr committed:
ci(travis): change config

kcmr committed:
ci(ember-try): add scenario for 3.4

kcmr committed:
docs(readme): remove hidden character

[skip ci]

kcmr committed:
fix(package): move demoURL to its corresponding field

kcmr committed:
chore(package): remove default included files and non existent folder

kcmr committed:
docs(readme): fix typos

kcmr committed:
refactor(index): change constant name

kcmr committed:
refactor(babel): remove duplicated code

josex2r commented on config/deploy.js:
```diff
@@ -0,0 +1,36 @@
+/* eslint-env node */
```
Why this addon has an `ember-cli-deploy` config file?

kcmr replied:
To publish the demo in gh-pages. The configuration is taken from https://github.com/BBVAEngineering/ember-cli-web-components

adrigzr replied:
This is used to deploy the dummy app into github gh-pages like in https://github.com/BBVAEngineering/ember-cli-web-components

Don't know if it is intended or a copy/paste thing

kcmr replied:
> Don't know if it is intended or a copy/paste thing

Both.

josex2r commented on index.js:
```diff
@@ -1,5 +1,127 @@
+		}
+	},
+
+	/* istanbul ignore next */
```
Why ignore? You can add a unit test to check that when calling this function it adds the `@webcomponents/webcomponentsjs` package.

kcmr replied:
Test included in 024e844

josex2r commented on .eslintignore:
```diff
@@ -12,6 +12,7 @@
 
 # misc
 /coverage/
+/utils/
```
This files should be linted, change the configuration for this directory from `node` to `browser`

adrigzr replied:
I think it is meant to be a polyfill. So it must go to vendor/public instead of utils.

kcmr replied:
Finally, the file is included in `lib/utils` as the `customElements.define` script will be always included in the bundle.

josex2r commented on index.js:
```diff
@@ -1,5 +1,127 @@
+		const filename = path.parse(this.options.outputFileName).name;
+		const ext = esmodule ? '-esm.js' : '.js';
+
+		return `${root}${this.options.outputPath}/${dirname}/${filename}${ext}`;
```
Always use `path.join(a, b, c)`

kcmr replied:
I think it's not necessary in this case because it's the `src` of the script tags.

josex2r commented on index.js:
```diff
@@ -1,5 +1,127 @@
+		this._super.included.apply(this, arguments);
+		this.app = parentAddon || app;
+
+		this._importPolyfills();
```
This addon can be loaded in async (see `_getScriptTag`) so, the polyfills should be added to the `vendor.js` file instead of defer its loading?

kcmr replied:
We can check if `autoImport` option is `false` to not include the polyfills in vendor, but this requires the user to know that `webcomponents-bundle.js` and `webcomponents-utils.js` (inside the addon) should be imported first... The second file is the problem here. Let me think about that.

kcmr replied:
Polyfills are not included if `autoImport` option is set to `false`.  Also the script for Safari 10 is removed (see the updated description of the PR).
Changed in a65a500

josex2r commented on node-tests/index-test.js:
```diff
@@ -0,0 +1,116 @@
+	);
+}
+
+function mockConfig(mockFile) {
```
Try not to use `sync` functions and use `async/await` instead , you disabled it at the first line of the line.

kcmr replied:
I think using `async/await` this way has the same result that using the sync versions of `fs` and it's a rule inherited from `eslint-plugin-node` that is ok not to follow for this kind of operations. I'm not making server requests and I really want file system operations to be synchronous. Also, using `async/await` requires at least Node 7.6 and the addon specifies that is supports version 6.* 

I'm changing it to not use sync methods, but I think that there is no improvement with this change and test files are allowed to not follow the [`no-sync` rule](https://eslint.org/docs/rules/no-sync).

kcmr replied:
Changed in b3f6c4c

adrigzr replied:
Node 6 has been deprecated, now we must use node 8 which has `async/await` native.

josex2r commented on node-tests/index-test.js:
```diff
@@ -0,0 +1,116 @@
+
+function mockConfig(mockFile) {
+	fs.renameSync(path.resolve(fixtureConfigPath, 'environment.js'), path.resolve(fixtureConfigPath, 'environment-BACKUP.js'));
+	fs.renameSync(mockFile, path.resolve(fixtureConfigPath, 'environment.js'));
```
Use vars to simplify the code:

```js
const envPath = resolve(fixtureConfigPath, 'environment.js');

function mockConfig(mockFile) {}
function restoreConfig(mockFile) {}
```

kcmr replied:
Done in 7adba59

josex2r commented on node-tests/index-test.js:
```diff
@@ -0,0 +1,116 @@
+		});
+
+		it('inserts the script tag for the bundle in index', () => {
+			assert.fileContent(outputFilePath('index.html'), '<script src="/assets/web-components/bundle.js"');
```
```js
const indexPath = outputFilePath('index.html');
```

kcmr replied:
✓

josex2r commented on node-tests/index-test.js:
```diff
@@ -0,0 +1,116 @@
+
+		after(() => {
+			restoreConfig(mockConfigFile);
+			fs.removeSync(distPath);
```
```js
await fs.remove(distPath);
```

kcmr replied:
✓

adrigzr commented on .nycrc.json:
```diff
@@ -0,0 +1,16 @@
+    "index.js",
+    "config/babel.js",
+    "config/output.js",
+    "config/rollup.js"
```
Add `config` instead?

kcmr replied:
There are some files inside `config/` that should not be included like `ember-try.js`, `environment.js` or `deploy.js`. Maybe I should use another folder for the addon specific files.

kcmr replied:
`lib/**` included instead of `config/`

adrigzr commented on .releaserc:
```diff
@@ -1,10 +1,20 @@
+  "plugins": [
+    [
+      "@semantic-release/exec", {
+        "publishCmd": "npx ember deploy production --verbose"
```
Ensure you have a gh-pages branch to deploy.

https://github.com/ef4/ember-cli-deploy-git#usage

kcmr replied:
Done. Thank you.

I think that the deploy config could be simplified by using `pages` as provider:

```
script: npm run build -- -prod

deploy:
  provider: pages
  skip_cleanup: true
  github-token: $GITHUB_TOKEN
  local-dir: dist
```

adrigzr replied:
Didn't know it can be done this way. Try it and we'll adopt it on following addons.

kcmr replied:
Done in f42757f. Not sure if this is going to work. Also `GITHUB_TOKEN` env variable needs to be set in Travis.

adrigzr commented on .travis.yml:
```diff
@@ -64,8 +65,12 @@ before_install:
 script:
-  - node_modules/.bin/ember try:one $EMBER_TRY_SCENARIO
+  - node_modules/.bin/ember try:one $EMBER_TRY_SCENARIO --skip_cleanup
+  - npm run test:node
```
This step must go into de "Tests" stage.

kcmr replied:
I've removed it because it's already included in `npm test` (executed in "Tests" stage).
Changed in dd51737

adrigzr commented on package.json:
```diff
@@ -1,47 +1,77 @@
+    "config/output.js",
+    "config/rollup.js",
+    "utils/**"
+  ],
```
Remove all unwanted files.

kcmr replied:
I'm using `files` field instead of `.npmignore`. It's easier to include only what you need in the package that to exclude a lot of config files and edit the `.npmignore` everytime you add a new unwanted file. You can check that the distributed package will be ok by executing `npm pack` inside the addon.

adrigzr replied:
I meant to remove `addon`, `app`, etc... folders from the project. You don't use them.

kcmr replied:
Ok. I removed some of them but not all...

kcmr replied:
Removed in f391834

adrigzr commented on package.json:
```diff
@@ -1,47 +1,77 @@
+    "test:node": "nyc mocha --recursive 'node-tests/**/*-test.js'",
+    "clean-coverage": "rm -rf coverage",
+    "merge-coverage": "cp coverage/ember/coverage-final.json coverage/ember.json && cp coverage/node/coverage-final.json coverage/node.json && nyc merge coverage coverage/coverage-final.json",
+    "check-coverage": "istanbul check-coverage coverage/coverage-final.json --statements 85 --functions 85 --branches 85 --lines 85",
```
100? 😄

kcmr replied:
If I use 100%, I'll have to test things like that:

```js
const options = baseConfig[this.name] || {};
env = env || process.env.EMBER_ENV;
```

Do you really think that this kind of test are useful?
For me they are a waste of time that only adds more noise to the tests.

adrigzr replied:
Then set it as the current coverage at least, in order to force check when coverage falls from the threshold.

kcmr replied:
Ok... 100% 9c01149

adrigzr replied:
🔝

adrigzr commented on package.json:
```diff
@@ -1,47 +1,77 @@
     "@semantic-release/changelog": "^3.0.4",
     "@semantic-release/git": "^7.0.12",
-    "broccoli-asset-rev": "^3.0.0",
+    "@webcomponents/webcomponentsjs": "^2.2.10",
```
I think this must be a production dependency...

kcmr replied:
I think that is a devDependency because it's used only for the demo, but I don't know if `addPackageToProject` requires this package to be included as an addon dependency.

kcmr replied:
After checking another [projects using `addPackageToProject`](https://github.com/search?q=this.addPackageToProject&type=Code), I found that is not required to have the dependencies that will be included in the hosting application listed in the addon package.json, so it's fine to have the dependency as devDependency.

adrigzr replied:
Nice.

adrigzr commented on package.json:
```diff
@@ -1,47 +1,77 @@
+    "ember-cli-deploy": "^1.0.2",
+    "ember-cli-deploy-build": "^1.1.1",
+    "ember-cli-deploy-git": "^1.3.3",
+    "ember-cli-deploy-git-ci": "^1.0.1",
```
Do you really need to deploy this addon?

kcmr replied:
Yeess, please :)

adrigzr commented on utils/webcomponents-utils.js:
```diff
@@ -0,0 +1,33 @@
+(function() {
```
Add some comment to know where did you get this file from to be able to update it in the future.

kcmr replied:
I don't know where this script comes from. I looked for the origins but I didn't found anything :/
It's frequently used when working with Web Components.

kcmr committed:
test(index): adds the required dependencies after installing

kcmr committed:
test(index): move integration tests to a folder

kcmr committed:
refactor(node-tests): use fs async methods in a sync manner

kcmr committed:
refactor(node-tests): move duplications to constants

kcmr committed:
refactor(config): move addon config files to a lib folder

kcmr committed:
chore(package): include lib folder in linted files

kcmr committed:
refactor(utils): move polyfill script to vendor folder

kcmr committed:
refactor(index): prevent always including polyfills in vendor

webcomponentsjs is not included in vendor when autoImport option is set to false

kcmr committed:
docs(readme): add a line break

kcmr committed:
chore(vendor): remove Babel optimizations

kcmr committed:
refactor: include customElements in the bundle and remove safari 10 fix

kcmr committed:
test: add unit tests and increase coverage thresholds

kcmr committed:
fix(package): ensure coverage folder is removed before merging files

kcmr committed:
fix(package): remove coverage folder before running tests

kcmr committed:
chore(package): remove non-existent folder from files field

kcmr committed:
ci(travis): remove duplicated step

npm test already executes test:node

adrigzr commented:
> I'm not very proud of stubbing "private" methods...

Me neither. It shouldn't be done.

kcmr committed:
chore: remove unused folders

kcmr committed:
style: fix lint warnings in .ember-cli

kcmr committed:
test(unit): prevent stubbing private methods

kcmr committed:
ci(deploy): use travis pages provider instead of ember-cli-deploy

kcmr commented:
> > I'm not very proud of stubbing "private" methods...
> 
> Me neither. It shouldn't be done.

Changed in 2f326d7 by mocking `addon.app`.

kcmr committed:
ci(travis): build dist before deploying

kcmr committed:
ci(travis): change env var name

kcmr committed:
fix(index): do not include polyfill if no entrypointPaths

kcmr committed:
fix(travis): bad indentation

adrigzr commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/ember-cli-webcomponents-bundler)
- [GitHub release](https://github.com/BBVAEngineering/ember-cli-webcomponents-bundler/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
